decompiler/pipeline/ssa/dependency_graph.py

```diff
@@ -1,74 +1,91 @@
-                    dependency_graph.add_edge(defined_variable, non_interfering_variable, strength="low")
-        return dependency_graph
-
-    def _non_interfering_requirements(self, requirements: List[Variable], defined_variable: Variable) -> Optional[Variable]:
-        """Get the unique non-interfering requirement if it exists, otherwise we return None."""
-        non_interfering_requirement = None
-        for required_variable in requirements:
-            if self._variables_can_have_same_name(defined_variable, required_variable):
-                if non_interfering_requirement:
-                    return None
-                non_interfering_requirement = required_variable
-        return non_interfering_requirement
-
-    def _variables_can_have_same_name(self, source: Variable, sink: Variable) -> bool:
-        """
-        Two variable can have the same name, if they have the same type, are both aliased or both non-aliased variables, and if they
-        do not interfere.
-
-        :param source: The potential source vertex.
-        :param sink: The potential sink vertex
-        :return: True, if the given variables can have the same name, and false otherwise.
-        """
-        if self.interference_graph.are_interfering(source, sink) or source.type != sink.type or source.is_aliased != sink.is_aliased:
-            return False
-        if source.is_aliased and sink.is_aliased and source.name != sink.name:
-            return False
-        return True
-
-    def get_components(self) -> Iterable[Set[Variable]]:
-        """Returns the weakly connected components of the dependency graph."""
-        for component in weakly_connected_components(self):
-            yield set(component)
+def _expression_dependencies(expression: Expression) -> dict[Variable, float]:
+    match expression:
+        case Variable():
+            return {expression: 1.0}
+        case Operation():
+            if expression.operation in {
+                OperationType.call,
+                OperationType.address,
+                OperationType.dereference,
+                OperationType.member_access,
+            }:
+                return {}
+
+            operands_dependencies = list(filter(lambda d: d, (_expression_dependencies(operand) for operand in expression.operands)))
+            dependencies: dict[Variable, float] = {}
+            for deps in operands_dependencies:
+                for var in deps:
+                    score = deps[var]
+                    score /= len(operands_dependencies)
+                    score *= 0.5  # penalize operations, so that expressions like (a + (a + (a + (a + a)))) gets a lower score than just (a)
+
+                    if var not in dependencies:
+                        dependencies[var] = score
+                    else:
+                        dependencies[var] += score
```

Comment:
I am still not sure whether this is what we want:

- for `expression=( a + 2*a) - a/4` has `a` the value 0.21875
- for `esxpression= a + b` has `a` the value 0.25

Shouldn't the first expression have a higher value for `a` than the second? We could simplify `( a + 2*a) - a/4` to `2.75 a` and the value would be 0.5.

Reply:
Perhaps I would change such that it has not the same "negative" influence than a binary operation. But I am not sure which value would be best. `0.9` sounds okay, since it rates nesting of a variable, but not too bad. Perhaps this could be either a defined-constant at the TOP to easier see that it could worth changing or a config option. Regarding the config option, I am not sure whether a user understands what this change influences, this is more helpful if we want to evaluate how different values change the output.

Reply:
Fixed with 8e7a65.